### ISSUE TITLE
Messages shown twice in Rainloop Webmail

### ISSUE THREAD

**Issue by jeefo** [foxcpp/maddy#251]
# Describe the bug

Messages in Rainloop Webmail is shown twice in all IMAP folders after some manipulations. It's working OK with dovecot, so posting as bug in maddy. If you delete message, both "copies" disappears, and appears in Trash wth same duplicate stuff. Afterlogic webmail is doing same. But Thunderbird / Outlook for Android, shows only one message.

Example: https://i.imgur.com/yRdvXqk.gif

# Steps to reproduce

Just view the mail in Webmail.

# Log files

Nothing wrong with logs.

# Configuration file

Default config from build.sh with changed hostname and sqlite db.

# Environment information

maddy 0.3.2+gf720dab linux/amd64 go1.14.3

default config: /etc/maddy/maddy.conf
default state_dir: /var/lib/maddy
default runtime_dir: /run/maddy

**Comment by foxcpp** [comment]
Thanks for the report. I will need dump of IMAP commands sent by Rainloop and exact responses generated by maddy. That could save me from the trouble of setting up Rainloop in a test environment :)

You need to add two directives into `imap` endpoint block like this:
```
imap ... {
  debug
  io_debug
  ...
}
```

**Comment by jeefo** [comment]
Here: https://gist.github.com/jeefo/cbc302eb4677b61d72661808162dc721

Moving from folder to folder, and tried to delete the msgs.

There was two msgs shown as four on start in the INBOX.

**Comment by foxcpp** [comment]
Okay, I also managed to reproduce the issue locally.

**Comment by foxcpp** [comment]
Aha, here is what seems to cause trouble:
```
TAG7 UID SEARCH UNDELETED
* SEARCH 1 1 2 2 4 4
TAG7 OK UID SEARCH completed
```

Now to figure out why UIDs are listed twice..

**Comment by foxcpp** [comment]
Issue source identified. One of SQL queries used to implement index returns duplicate rows due to missing `DISTINCT`.

**Comment by foxcpp** [comment]
Fix commit: https://github.com/foxcpp/maddy/commit/aa2bcf13528a91178785b897f18b1fe99e8abfdd

**Comment by foxcpp** [comment]
Will include in 0.3.3 release.

**Comment by jeefo** [comment]
Thank you. 

Another problem is that Attachments are all shown as 0 bytes, but downloadable OK. Not sure if this deserve new issue report.

Not only in RainLoop, but in desktop clients too.

**Comment by foxcpp** [comment]
>Another problem is that Attachments are all shown as 0 bytes, but downloadable OK. Not sure if this deserve new issue report.
>Not only in RainLoop, but in desktop clients too.

This is caused by incomplete implementation that does not calculate "body part" size information.  Oh, I actually sent a PR to go-imap that fixes it and it was merged 27 days ago. I forgot to update version used in maddy, thanks for the reminder! :)

**Comment by foxcpp** [comment]
I am currently waiting for one more fix to be merged in go-smtp upstream before tagging 0.3.3.

You can get the maddy build with other fixes that will go into 0.3.3 by checking out 0.3-fixes branch in this repo:
```
git clone https://github.com/foxcpp/maddy.git
cd maddy
git checkout 0.3-fixes
./build.sh
```

**Comment by jeefo** [comment]
Everything is fine! Thank you.
